Changelog for 8.20 (#19281)
coq/coq: doc/sphinx/changes.rst

   .. include:: ../unreleased.rst

Version 8.20
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

Coq version 8.20 TODO_main_new_features

We highlight some of the most impactful changes here:

  - :ref:`rewrite_rules`

  - :ref:`primitive strings <pstring_library>`

>   - TODO other notable changes

Reviewer: A lot of work went into reducing the size of the bytecode segment, which in turn means that `.vo` files might now be considerably smaller.

  by Jan-Oliver Kaiser and Rodolphe Lepigre).
- **Changed:**
  Tactic :g:`intro z` on an existential variable goal forces the resolution
  of the existential variable into a goal :g:`forall z:?T, ?P`, which
  becomes :g:`?P` in context :g:`z:?T` after introduction. The
  existential variable :n:`?P` itself is now defined in a context
  where the variable of type `?T` is also named :g:`z`, as specified
  by :tacn:`intro` instead of :g:`x` as it was conventionally the case
  before
  (`#18395 <https://github.com/coq/coq/pull/18395>`_,
  by Hugo Herbelin).
- **Changed:**
  syntactic global references passed through the `using` clauses of :tacn:`auto`-like
  tactics are now handled as plain references rather than interpreted terms. In
  particular, their typeclass arguments will not be inferred. In general, the previous
  behaviour can be emulated by replacing `auto using foo` with `pose proof foo; auto`
  (`#18909 <https://github.com/coq/coq/pull/18909>`_,
  by Pierre-Marie Pédrot).
- **Changed:**
  Use Coqlib's Register mechanism for the generalized rewriting tactic
  and make the (C)RelationClasses/(C)Morphisms independent of the `rewrite`
  tactic to ease maintainance.

>   (`#19115 <https://github.com/coq/coq/pull/19116>`_,

Reviewer: ```suggestion
  (`#19115 <https://github.com/coq/coq/pull/19115>`_,
```

Gallego Arias, Gaëtan Gilbert, Georges Gonthier, Benjamin Grégoire,
Jason Gross, Hugo Herbelin, Vincent Laporte, Olivier Laurent, Assia
Mahboubi, Kenji Maillard, Guillaume Melquiond, Pierre-Marie Pédrot,
Clément Pit-Claudel, Pierre Roux, Kazuhiko Sakaguchi, Vincent Semeria,
Michael Soegtrop, Arnaud Spiwack, Matthieu Sozeau, Enrico Tassi,
Laurent Théry, Anton Trunov, Li-yao Xia and Théo Zimmermann. See the
`Coq Team face book <https://coq.inria.fr/coq-team.html>`_ page for
more details. TODO_update

Assignees: 16
Villetaneuse, andres-erbsen, fajb, Ali Caglayan, Jim Fehrle, Gaëtan Gilbert, Jason Gross, Hugo Herbelin, Emilio Jesús Gallego Arias, Guillaume Melquiond, Pierre-Marie Pédrot, Pierre Roux, Michael Soegtrop, Matthieu Sozeau, Enrico Tassi, Théo Zimmermann,

Reviewers: 32
Alizter, CohenCyril, Janno, JasonGross, LasseBlaauwbroek, MSoegtropIMC, RalfJung, SkySkimmer, Villetaneuse, Zimmi48, andres-erbsen, ckeller, cpitclaudel, ejgallego, fajb, gares, herbelin, jfehrle, liyishuai, mattam82, mrhaandi, nmvdw, olaure01, palmskog, ppedrot, proux01, rlepigre, rmatthes, silene, soukouki, tabareau, zeldovich,

The 54 contributors to the 8.20 version are:

> Arya-Elfren, Remyjck, chluebi, doctor-kaliy, lduboisd, qvermande,  TODO_new_contributors_to_precise

Author: Quentin already appears below
```suggestion
Arya-Elfren, Remyjck, chluebi, doctor-kaliy, lduboisd,  TODO_new_contributors_to_precise
```

Timur Aminev, Frédéric Besson, Lasse Blaauwbroek, Björn Brandenburg,

> Ali Caglayan, Nikolaos Chatzikonstantinou, Sylvain Chiron, Cyril

Author: @Arya-Elfren under what name do you want to appear in the forthcoming 8.20 changelog?

Reviewer 2: `Oliver Turner` would be best. Thanks!

> Ali Caglayan, Nikolaos Chatzikonstantinou, Sylvain Chiron, Cyril

Author: @Remyjck shall we list you as `Remy Seassau` in the forthcoming 8.20 changelog?

Author: Done, @Remyjck please complain here (before next week) if you prefer something else.

> Ali Caglayan, Nikolaos Chatzikonstantinou, Sylvain Chiron, Cyril

Author: @chluebi under what name do you want to appear in the forthcoming 8.20 changelog?

Reviewer 3: chluebi

Author: Thanks for your answer (note that considering the list mostly contains plain, capitalized, names, there is a risk we ask the question for every new release you contribute to).

> Ali Caglayan, Nikolaos Chatzikonstantinou, Sylvain Chiron, Cyril

Author: @doctor-kaliy shall we list you as `Evgenii Kosogorov` in the forthcoming 8.20 changelog?

Reviewer 4: Yes please! Thank you!

> Ali Caglayan, Nikolaos Chatzikonstantinou, Sylvain Chiron, Cyril

Author: @langston-barrett shall we list you as `Langston Barrett` in the forthcoming 8.20 changelog?

Author: Done, @louiseddp please complain here (before next week) if you prefer something else.

Reviewer 5: Hello @proux01, I just saw your message ! Using my full name `Louise Dubois de Prisque` is perfect. Thank you very much !

  * Fix double printing after `#go` by making sure that the feeder is added in a separate function that is only executed once
  * Make the "include" file run automatically, without asking users to do it manually or adding anything to `~/.ocamlinit`
  * Clean up the include files, move them to a separate directory `dev/ml_toplevel`. Fix all the errors and warnings in them
  * Fix the Dune files as to copy the include files into the `_build` directory
  * Add a custom OCaml toplevel directory `#go` which does the same action as `go ()`, but with a more native syntax
  * Make the include files not modify the global flags permenantly
  * Rename `Flags.in_toplevel` to `Flags.in_ml_toplevel` to highlight the fact that it is about the OCaml toplevel and not the Coq one
  * Rename `dev/shim/coqbyte` to `dev/shim/coqtop.byte` to match the standard installation convention and make it easier to find
  * Update the debugging documentation, fix formatting there
  * Add a new `coq_config_bin.ml` file, and also unify how coqtop and coqc are built
  (`#18771 <https://github.com/coq/coq/pull/18771>`_,

>   by Anton Danilkin).

Author: TODO: cleanup keeping only user visible changes

  Emilio Jesus Gallego Arias).
- **Added:**
  Coq is now compatible with `memprof-limits` interruption
  methods. This means that Coq will be recompiled when the library is
  installed / removed from an OPAM switch.
  (`#18906 <https://github.com/coq/coq/pull/18906>`_,
  fixes `#17760 <https://github.com/coq/coq/issues/17760>`_,
  by Emilio Jesus Gallego Arias).

Extraction
^^^^^^^^^^

- **Added:**
  Extension for OCaml extraction:
  Commands to extract foreign function calls to C (external)
  and ML function exposition (Callback.register) for calling
  being able to call them by C functions
  (`#18270 <https://github.com/coq/coq/pull/18270>`_,
  fixes `#18212 <https://github.com/coq/coq/issues/18212>`_,
  by Mario Frank).

Miscellaneous
^^^^^^^^^^^^^

- **Fixed:**
  Wrongly self-referencing extraction of primitive projections to OCaml in functors
  (`#17321 <https://github.com/coq/coq/pull/17321>`_,
  fixes `#16288 <https://github.com/coq/coq/issues/16288>`_,
  by Hugo Herbelin). Note that OCaml wrappers assuming that the
  applicative syntax of projections is provided may have
  to use the dot notation instead.

Version 8.19

> ------------

Reviewer 6: This isolated Miscellaneous item could actually be moved to Extraction.

Also, some items start with a lowercase and others with an uppercase. I don't know which is best.

Author: There are two way to see the Changed/Added/Fixed/...
* either as the first word of the sentence (hence no second capitalization)
* as a title (hence capitalization just after)

If we don't enforce one of the two styles, we'll keep having both (not a major issue IMHO).

   .. include:: ../unreleased.rst

Version 8.20
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

Coq version 8.20 adds a new rewrite rule mechanism along with a few
new features, a host of improvements to the virtual machine, the
notation system, Ltac2 and the standard library.

We highlight some of the most impactful changes here:

  - :ref:`rewrite_rules`

  - :ref:`primitive strings <pstring_library>`

  - A lot of work went into reducing the size of the bytecode segment,
    which in turn means that .vo files might now be considerably
    smaller.

  - A new version of the Docker keeper tool to build and maintain

>     Docker images of Coq.

Author: @Zimmi48 @erikmd could you please check that sentence and provide a link to point to?

    does not undo its mutations. Replace `Ltac2 mutable foo :=
    some_expr.` with `Ltac2 mutable foo := some_expr. Ltac2 Set foo :=
    some_expr.` to recover the previous behaviour.

  - Some :ref:`renaming <820_renaming_stdlib>` in the standard
    library. Deprecations are provided for a smooth transition.

See the `Changes in 8.20.0`_ section below for the detailed list of changes,
including potentially breaking changes marked with **Changed**.
Coq's `reference manual for 8.20 <https://coq.github.io/doc/v8.20/refman>`_,
`documentation of the 8.20 standard library <https://coq.github.io/doc/v8.20/stdlib>`_
and `developer documentation of the 8.20 ML API <https://coq.github.io/doc/v8.20/api>`_
are also available.

Théo Zimmermann with help from Ali Caglayan and Jason Gross maintained
`coqbot <https://github.com/coq/bot>`_ used to run Coq's CI and other
pull request management tasks.

Jason Gross maintained the `bug minimizer <https://github.com/JasonGross/coq-tools>`_
and its `automatic use through coqbot <https://github.com/coq/coq/wiki/Coqbot-minimize-feature>`_.

Jaime Arias and Erik Martin-Dorel maintained the
`Coq Docker images <https://hub.docker.com/r/coqorg/coq>`_

> and the Docker Keeper tool used to build and keep those images up to date.

Author: @Zimmi48 @erikmd could you please check that sentence

Reviewer 7: Thanks @proux01! I'm the sole author/maintainer of docker-keeper, and we may want to highlight that the tool is generic (it is a compiler from a YAML DSL to GitLab-CI child pipelines, also serialized in YAML).

I propose the following sentence, let me know if it looks good to you:
```suggestion
Jaime Arias and Erik Martin-Dorel maintained the
`Coq Docker images <https://hub.docker.com/r/coqorg/coq>`_.
Erik Martin-Dorel maintained the
`docker-keeper <https://gitlab.com/erikmd/docker-keeper>`_ compiler,
used to build and keep those images up to date, which is generic and applicable to projects beyond Coq.
```

Author: What about " (note that the tool is not Coq specific)"?
If you prefer compiler to tool, should we also update the other comment?

Reviewer 7: > What about " (note that the tool is not Coq specific)"?

Yes if you want, if you think it's better.

> If you prefer compiler to tool, should we also update the other comment?

Yes, good point :) it'd be more consistent then.

Cyril Cohen, Vincent Laporte, Pierre Roux and Théo Zimmermann
maintained the `Nix toolbox <https://github.com/coq-community/coq-nix-toolbox>`_
used by many Coq projects for continuous integration.

Ali Caglayan, Emilio Jesús Gallego Arias, Rudi Grinberg and
Rodolphe Lepigre maintained the
`Dune build system for OCaml and Coq <https://github.com/ocaml/dune/>`_
used to build Coq itself and many Coq projects.

The opam repository for Coq packages has been maintained by
Guillaume Claret, Guillaume Melquiond, Karl Palmskog and Enrico Tassi with
contributions from many users. A list of packages is `available on the Coq website <https://coq.inria.fr/coq-package-index>`_.

Coq 8.20 was made possible thanks to the following reviewers:
Frédéric Besson, Lasse Blaauwbroek, Ali Caglayan, Cyril Cohen, Andrej
Dudenhefner, Andres Erbsen, Jim Fehrle, Emilio Jesús Gallego Arias,
Gaëtan Gilbert, Jason Gross, Hugo Herbelin, Ralf Jung, Jan-Oliver
Kaiser, Chantal Keller, Olivier Laurent, Rodolphe Lepigre, Yishuai Li,
Ralph Matthes, Guillaume Melquiond, Pierre-Marie Pédrot, Karl
Palmskog, Clément Pit-Claudel, Pierre Rousselin, Pierre Roux, Michael

> Soegtrop, soukouki, Matthieu Sozeau, Nicolas Tabareau, Enrico Tassi,

Author: @soukouki you're welcome to provide a full name if you prefer (before next Wednesday, sorry for the short notice)

   .. include:: ../unreleased.rst

Version 8.20
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

Coq version 8.20 TODO_main_new_features

We highlight some of the most impactful changes here:

  - :ref:`rewrite_rules`

  - :ref:`primitive strings <pstring_library>`

  - A lot of work went into reducing the size of the bytecode segment,
    which in turn means that .vo files might now be considerably
    smaller.

>   - TODO other notable changes

Author: add mention of new version of Docker keeper

Notable breaking changes:

  - TODO I can identify

>   - syntactic global references passed through the `using` clauses of

Author: Not so sure about that one, there is a link to about the same thing three words later and I feel `using` is more readable than `auto_using`, because it's the user visible syntax, rather than some internal parser name.

>   - syntactic global references passed through the `using` clauses of

Reviewer 8: `auto_using` appears in the doc in the syntax of (I think) all the relevant commands (listed below), so it has a nice precision to it.  Maybe say ``` `using` clauses (:n:`@auto_using`) ```?

```
| "trivial" OPT auto_using OPT hintbases
| "info_trivial" OPT auto_using OPT hintbases
| "debug" "trivial" OPT auto_using OPT hintbases
| "auto" OPT nat_or_var OPT auto_using OPT hintbases
| "info_auto" OPT nat_or_var OPT auto_using OPT hintbases
| "debug" "auto" OPT nat_or_var OPT auto_using OPT hintbases
| "eauto" OPT nat_or_var OPT auto_using OPT hintbases
| "debug" "eauto" OPT nat_or_var OPT auto_using OPT hintbases
| "info_eauto" OPT nat_or_var OPT auto_using OPT hintbases
| "dfs" "eauto" OPT nat_or_var OPT auto_using OPT hintbases
```

>   - syntactic global references passed through the `using` clauses of

Author: Still feels rather redundant to me (since there is already a link to auto, auto_using is just one more click away anyway) but any stronger rational / opinion from anyone else is welcome.

  - :cmd:`Import`\ing a module containing a mutable Ltac2 definition
    does not undo its mutations. Replace `Ltac2 mutable foo :=
    some_expr.` with `Ltac2 mutable foo := some_expr. Ltac2 Set foo :=
    some_expr.` to recover the previous behaviour

  - TODO other breaking changes to mention here?

See the `Changes in 8.20.0`_ section below for the detailed list of changes,
including potentially breaking changes marked with **Changed**.
Coq's `reference manual for 8.20 <https://coq.github.io/doc/v8.20/refman>`_,
`documentation of the 8.20 standard library <https://coq.github.io/doc/v8.20/stdlib>`_
and `developer documentation of the 8.20 ML API <https://coq.github.io/doc/v8.20/api>`_
are also available.

Théo Zimmermann with help from Ali Caglayan and Jason Gross maintained
`coqbot <https://github.com/coq/bot>`_ used to run Coq's CI and other
pull request management tasks. TODO_check_paragraph

Jason Gross maintained the `bug minimizer <https://github.com/JasonGross/coq-tools>`_
and its `automatic use through coqbot <https://github.com/coq/coq/wiki/Coqbot-minimize-feature>`_.

Jaime Arias and Erik Martin-Dorel maintained the

> `Coq Docker images <https://hub.docker.com/r/coqorg/coq>`_

Author: and docker keeper

contributions from many users. A list of packages is `available on the Coq website <https://coq.inria.fr/coq-package-index>`_.

Our current maintainers are Yves Bertot, Frédéric Besson, Ana Borges,
Ali Caglayan, Tej Chajed, Cyril Cohen, Pierre Corbineau, Pierre
Courtieu, Andres Erbsen, Jim Fehrle, Julien Forest, Emilio Jesús
Gallego Arias, Gaëtan Gilbert, Georges Gonthier, Benjamin Grégoire,
Jason Gross, Hugo Herbelin, Vincent Laporte, Olivier Laurent, Assia
Mahboubi, Kenji Maillard, Guillaume Melquiond, Pierre-Marie Pédrot,
Clément Pit-Claudel, Pierre Roux, Kazuhiko Sakaguchi, Vincent Semeria,
Michael Soegtrop, Arnaud Spiwack, Matthieu Sozeau, Enrico Tassi,
Laurent Théry, Anton Trunov, Li-yao Xia and Théo Zimmermann. See the
`Coq Team face book <https://coq.inria.fr/coq-team.html>`_ page for
more details. TODO_update

Assignees: 16
Frédéric Besson, Ali Caglayan, Andres Erbsen, Jim Fehrle, Gaëtan Gilbert, Jason Gross, Hugo Herbelin, Emilio Jesús Gallego Arias, Guillaume Melquiond, Pierre-Marie Pédrot, Pierre Rousselin, Pierre Roux, Michael Soegtrop, Matthieu Sozeau, Enrico Tassi, Théo Zimmermann,

Reviewers: 32

> Alizter, CohenCyril, Janno, JasonGross, LasseBlaauwbroek, MSoegtropIMC, RalfJung, SkySkimmer, Villetaneuse, Zimmi48, andres-erbsen, ckeller, cpitclaudel, ejgallego, fajb, gares, herbelin, jfehrle, liyishuai, mattam82, mrhaandi, nmvdw, olaure01, palmskog, ppedrot, proux01, rlepigre, rmatthes, silene, soukouki, tabareau, zeldovich,

Author: Sorry, I should have kept some TODO comment, that was indeed a temporary draft.

  or, using :g:`Import Z.euclidean_division_equations_flags.`, with
  :g:`Z.to_euclidean_division_equations_with ltac:(default_with find_duplicate_quotients false)`
  (`#17934 <https://github.com/coq/coq/pull/17934>`_,
  by Jason Gross).
- **Changed:**
  The opacity/transparency of primitive projections is now attached to the
  projections themselves, not the compatibility constants, and compatibility
  constants are always considered transparent
  (`#18327 <https://github.com/coq/coq/pull/18327>`_,
  fixes `#18281 <https://github.com/coq/coq/issues/18281>`_,
  by Jan-Oliver Kaiser and Rodolphe Lepigre).
- **Changed:**
  Tactic :g:`intro z` on an existential variable goal forces the resolution
  of the existential variable into a goal :g:`forall z:?T, ?P`, which
  becomes :g:`?P` in context :g:`z:?T` after introduction. The
  existential variable :n:`?P` itself is now defined in a context
  where the variable of type `?T` is also named :g:`z`, as specified
  by :tacn:`intro` instead of :g:`x` as it was conventionally the case
  before
  (`#18395 <https://github.com/coq/coq/pull/18395>`_,
  by Hugo Herbelin).
- **Changed:**

>   syntactic global references passed through the `using` clauses of :tacn:`auto`-like

Author: C.f. https://github.com/coq/coq/pull/19281#discussion_r1734271854

  tactics are now handled as plain references rather than interpreted terms. In
  particular, their typeclass arguments will not be inferred. In general, the previous
  behaviour can be emulated by replacing `auto using foo` with `pose proof foo; auto`
  (`#18909 <https://github.com/coq/coq/pull/18909>`_,
  by Pierre-Marie Pédrot).
- **Changed:**

>   Use Coqlib's Register mechanism for the generalized rewriting tactic

Reviewer 8: ```suggestion
  Use Coqlib's :cmd:`Register` mechanism for the generalized rewriting tactic
```

IIUC

  and make the (C)RelationClasses/(C)Morphisms independent of the `rewrite`
  tactic to ease maintainance.
  (`#19115 <https://github.com/coq/coq/pull/19115>`_,
  by Matthieu Sozeau).
- **Removed:**

>   the clear modifier which was deprecated since 8.17

Reviewer 8: ```suggestion
  the `clear` modifier which was deprecated since 8.17
```

  (`#18887 <https://github.com/coq/coq/pull/18887>`_,
  by Pierre-Marie Pédrot).
- **Removed:**
  the `cutrewrite` tactic, which was deprecated since
  Coq 8.5
  (`#19027 <https://github.com/coq/coq/pull/19027>`_,
  by Pierre-Marie Pédrot).
- **Deprecated:**

>   non-reference hints in `using` clauses of :tacn:`auto`-like tactics

Reviewer 8: ```suggestion
  non-reference hints in :n:`@auto_using` clauses of :tacn:`auto`-like tactics
```

Author: C.f., https://github.com/coq/coq/pull/19281#discussion_r1734271854

  (`#19006 <https://github.com/coq/coq/pull/19006>`_,
  by Pierre-Marie Pédrot).
- **Deprecated:**
  the :tacn:`gintuition` tactic, which used to be undocumented
  until Coq 8.16
  (`#19129 <https://github.com/coq/coq/pull/19129>`_,
  by Pierre-Marie Pédrot).
- **Deprecated:**

>   :tacn:`destauto`.

Reviewer 8: ```suggestion
  :tacn:`destauto`
```

Author: Not so sure since there is a capital on next line, uncapitalized it and replaced dot by a comma

  like it does for :tacn:`cbn`, :tacn:`lazy` and :tacn:`simpl`
  (`#18618 <https://github.com/coq/coq/pull/18618>`_,
  fixes `#9086 <https://github.com/coq/coq/issues/9086>`_,
  by Hugo Herbelin).

Ltac language
^^^^^^^^^^^^^
- **Added:**
  In :tacn:`rewrite_strat`, :n:`@rewstrategy` now supports the fixpoint operator :n:`fix @ident := @rewstrategy1`
  (`#18094 <https://github.com/coq/coq/pull/18094>`_,
  fixes `#13702 <https://github.com/coq/coq/issues/13702>`_,
  by Jason Gross and Gaëtan Gilbert).
- **Fixed:**
  :tacn:`rewrite_strat` now works inside module functors
  (`#18094 <https://github.com/coq/coq/pull/18094>`_,
  fixes `#18463 <https://github.com/coq/coq/issues/18463>`_,
  by Jason Gross).

Ltac2 language
^^^^^^^^^^^^^^
- **Changed:**
  recursive `let` and non mutable projections of syntactic values are considered syntactic values
  (`#18411 <https://github.com/coq/coq/pull/18411>`_,
  by Gaëtan Gilbert).
- **Changed:**
  Ltac2 are typechecked at declaration time by default.
  This should produce better errors when a notation argument does not have the expected type
  (e.g. wrong branch type in `match! goal`).

>   The previous behaviour of typechecking only the expansion result can be

Reviewer 8: ```suggestion
  In the previous behaviour of typechecking, only the expansion result can be
```

  recovered using :flag:`Ltac2 Typed Notations`. We believe there are no real
  use cases for this, please report if you have any
  (`#18432 <https://github.com/coq/coq/pull/18432>`_,
  fixes `#17477 <https://github.com/coq/coq/issues/17477>`_,
  by Gaëtan Gilbert).
- **Changed:**
  argument order for the Ltac2 combinators `List.fold_left2` and `List.fold_right2`
  changed to be the same as in OCaml
  (`#18706 <https://github.com/coq/coq/pull/18706>`_,
  by Gaëtan Gilbert).
- **Changed:**
  :cmd:`Import`\ing a module containing a mutable Ltac2 definition
  does not undo its mutations. Replace `Ltac2 mutable foo := some_expr.`
  with `Ltac2 mutable foo := some_expr. Ltac2 Set foo := some_expr.`
  to recover the previous behaviour
  (`#18713 <https://github.com/coq/coq/pull/18713>`_,
  by Gaëtan Gilbert).
- **Changed:**

>   the using clause argument of :tacn:`auto`-like tactics in Ltac2 now

Reviewer 8: ```suggestion
  the :n:`@auto_using` clause of :tacn:`auto`-like tactics in Ltac2 now
```

Author: C.f., https://github.com/coq/coq/pull/19281#discussion_r1734271854
